nodestream/project/pipeline_definition.py

```diff
@@ -66,38 +95,66 @@ def describe_yaml_schema(cls):
         )
 
     @classmethod
-    def from_file_data(cls, data, parent_annotations):
+    def from_file_data(
+        cls, data, parent_configuration: PipelineConfiguration
+    ) -> "PipelineDefinition":
         if isinstance(data, str):
             data = {"path": data}
 
         file_path = Path(data.pop("path"))
         name = data.pop("name", get_default_name(file_path))
-        exclude_targets = data.pop("exclude_inherited_targets", False)
-        annotations = data.pop("annotations", {})
-        targets = data.pop("targets", [])
-        return cls(
-            name,
-            file_path,
-            set(targets),
-            exclude_targets,
-            {**parent_annotations, **annotations},
-        )
+
+        configuration = PipelineConfiguration.from_file_data(data)
+        configuration.parent = parent_configuration
+        return cls(name, file_path, configuration)
+
+    @classmethod
+    def from_plugin_data(
+        cls, data, parent_configuration: PipelineConfiguration
+    ) -> "PipelineDefinition":
+        name = data.pop("name")
+        configuration = PipelineConfiguration.from_file_data(data)
+        configuration.parent = parent_configuration
+        return cls(name, None, configuration)
 
     def to_file_data(self, verbose: bool = False):
         using_default_name = self.name == self.file_path.stem
-        if using_default_name and not self.annotations and not verbose:
+        if using_default_name and not self.configuration.annotations and not verbose:
             return str(self.file_path)
 
         result = {"path": str(self.file_path)}
         if not using_default_name or verbose:
             result["name"] = self.name
-        if self.annotations or verbose:
-            result["annotations"] = self.annotations
-        if self.targets or verbose:
-            result["targets"] = self.targets
+        annotations = self.configuration.effective_annotations
+        targets = self.configuration.effective_targets
+        if targets or verbose:
+            result["targets"] = list(targets)
+        if annotations or verbose:
+            result["annotations"] = annotations
+        if self.configuration.exclude_inherited_targets or verbose:
+            result[
+                "exclude_inherited_targets"
+            ] = self.configuration.exclude_inherited_targets
```

Comment:
This is duplicated with code in `to_plugin_file_data`. It also has feature envy on `PipelineConfiguration`. Should this code become `PipelineConfiguration#to_file_data`